EIP-7547 (#3617)
ethereum/consensus-specs: specs/_features/eip7547/p2p-networking.md

- [EIP-7547 -- Networking](#eip-7547----networking)
  - [Containers](#containers)
    - [New Containers](#new-containers)
      - [`SignedInclusionList`](#signedinclusionlist)
  - [Modifications in EIP7547](#modifications-in-eip7547)
    - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
      - [Topics and messages](#topics-and-messages)
        - [Global topics](#global-topics)
          - [New `inclusion_list`](#new-inclusion_list)
      - [Transitioning the gossip](#transitioning-the-gossip)
  - [Design rationale](#design-rationale)
  - [Why is it proposer may send multiple inclusion lists? Why not just one per slot?](#why-is-it-proposer-may-send-multiple-inclusion-lists-why-not-just-one-per-slot)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# EIP-7547 -- Networking

This document contains the consensus-layer networking specification for EIP-7547.

The specification of these changes continues in the same format as the network specifications of previous upgrades, and assumes them as pre-requisite.

## Containers

### New Containers

#### `SignedInclusionList`

```python
class SignedInclusionList(Container):

>     message: InclusionList

Author: no strong opinion on this, will ask potuz

    signature: BLSSignature
```

## Modifications in EIP7547

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork for EIP7547 to support upgraded types.

#### Topics and messages

Topics follow the same specification as in prior upgrades.

The derivation of the `message-id` remains stable.

##### Global topics

###### New `inclusion_list`

The *type* of the payload of this topic is `SignedInclusionList`.

New validation:

The following validations MUST pass before forwarding the inclusion_list on the network.

> - _[REJECT]_ The slot `message.signedSummary.message.slot` is greater than or equal to the current slot (clients can queue future inclusion lists).

Author: for sure!

> - _[REJECT]_ The slot `message.signedSummary.message.slot` is greater than or equal to the current slot (clients can queue future inclusion lists).

Author: done

  - [Why is it proposer may send multiple inclusion lists? Why not just one per slot?](#why-is-it-proposer-may-send-multiple-inclusion-lists-why-not-just-one-per-slot)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# EIP-7547 -- Networking

This document contains the consensus-layer networking specification for EIP-7547.

The specification of these changes continues in the same format as the network specifications of previous upgrades, and assumes them as pre-requisite.

## Containers

### New Containers

#### `SignedInclusionList`

```python
class SignedInclusionList(Container):
    message: InclusionList
    signature: BLSSignature
```

## Modifications in EIP7547

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork for EIP7547 to support upgraded types.

#### Topics and messages

Topics follow the same specification as in prior upgrades.

The derivation of the `message-id` remains stable.

##### Global topics

###### New `inclusion_list`

The *type* of the payload of this topic is `SignedInclusionList`.

New validation:

The following validations MUST pass before forwarding the inclusion_list on the network.

> - _[REJECT]_ The slot `message.signedSummary.message.slot` is the current slot.

Reviewer: We should make a note `a client MAY queue future IL for processing at the appropriate slot`

- _[REJECT]_ The `signature` is valid for the current proposer `message.signedSummary.message.proposer_index`.

> - _[REJECT]_ The `message.signedSummary.message.parent_hash` is not the current head.

Reviewer: The parent hash refers to an EL block, whereas when we mention "head," we typically refer to a consensus block. I'm also unsure about the purpose of this check : )

Author: removed!

> - _[REJECT]_ The inclusion list transactions `message.transactions` length is within upperbound `MAX_TRANSACTIONS_PER_INCLUSION_LIST`.

Reviewer: This feels more like an EL responsibility

Author: ill remove the parent hash one. i think the other stuff feels like CL side verification, no ?

      - [`SignedBeaconBlockAndInclusionList`](#signedbeaconblockandinclusionlist)
  - [Modifications in EIP7547](#modifications-in-eip7547)
    - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
      - [Topics and messages](#topics-and-messages)
        - [Global topics](#global-topics)
          - [`beacon_block`](#beacon_block)
      - [Transitioning the gossip](#transitioning-the-gossip)
  - [Design rationale](#design-rationale)
  - [Why is it proposer may send multiple inclusion lists? Why not just one per slot?](#why-is-it-proposer-may-send-multiple-inclusion-lists-why-not-just-one-per-slot)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# EIP-7547 -- Networking

This document contains the consensus-layer networking specification for EIP-7547.

The specification of these changes continues in the same format as the network specifications of previous upgrades, and assumes them as pre-requisite.

## Preset

### Execution

| Name | Value |
| - | - |
| `MAX_TRANSACTIONS_PER_INCLUSION_LIST` |  `uint64(143)` |

## Containers

### New Containers

#### `SignedInclusionList`

```python
class SignedInclusionList(Container):

>     signed_summary: SignedInclusionListSummary

Reviewer 2: can we have the slot (at which this is emitted)

>     signed_summary: SignedInclusionListSummary

Reviewer 2: can we have slot here or in the signed summary?

Author: done!

The specification of these changes continues in the same format as the network specifications of previous upgrades, and assumes them as pre-requisite.

## Preset

### Execution

| Name | Value |
| - | - |
| `MAX_TRANSACTIONS_PER_INCLUSION_LIST` |  `uint64(143)` |

## Containers

### New Containers

#### `SignedInclusionList`

```python
class SignedInclusionList(Container):
    signed_summary: SignedInclusionListSummary
    transactions: List[Transaction, MAX_TRANSACTIONS_PER_INCLUSION_LIST]
    signature: BLSSignature
```

#### `SignedBeaconBlockAndInclusionList`

```python
class SignedBeaconBlockAndInclusionList(Container):
    signed_block: SignedBeaconBlock
    signed_inclusion_list: SignedInclusionList
```

## Modifications in EIP7547

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork for EIP7547 to support upgraded types.

#### Topics and messages

Topics follow the same specification as in prior upgrades.

The derivation of the `message-id` remains stable.

##### Global topics

###### `beacon_block`

The *type* of the payload of this topic changes to the (modified) `SignedBeaconBlockAndInclusionList`.

New validation:

> The following validations MUST pass before forwarding the `beacon_block` on the network.

Reviewer: ```suggestion
The following validations MUST pass before forwarding the signed_beacon_block_and_inclusion_list on the network. (We define the following for convenience -- signed_block = signed_beacon_block_and_inclusion_list.signed_block and signed_inclusion_list = signed_beacon_block_and_inclusion_list.signed_inclusion_list)
```

> - _[REJECT]_ The inclusion list transactions `beacon_block.transactions` length is within upperbound `MAX_TRANSACTIONS_PER_INCLUSION_LIST`.

Reviewer: ```suggestion
- _[REJECT]_ The inclusion list transactions `signed_inclusion_list.transactions` length is within upperbound `MAX_TRANSACTIONS_PER_INCLUSION_LIST`.
```

> - _[REJECT]_ The inclusion list summary has the same length of transactions `len(beacon_block.signed_block.inclusion_list_summary) == len(beacon_block.transactions)`.

Reviewer: ```suggestion
- _[REJECT]_ The inclusion list summary has the same length of transactions `len(signed_inclusion_list.signed_summary.summary) == len(signed_inclusion_list.transactions)`.
```

- _[REJECT]_ The inclusion list transactions signature, `beacon_block.signed_transactions.signature`, is valid with respect to the `proposer_index` pubkey.

> - _[REJECT]_ The summary is proposed by the expected proposer_index for the summary's slot in the context of the current shuffling (defined by parent_root/slot). If the proposer_index cannot immediately be verified against the expected shuffling, the inclusion list MAY be queued for later processing while proposers for the summary's branch are calculated -- in such a case do not REJECT, instead IGNORE this message.

Reviewer: No longer needed if it's coupled 

```suggestion
```
